src/polymer/polymer2-element-scanner.ts

```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
+          jsdoc.hasTag(doc, 'polymerElement')) {
+        this._classFound(assignedName, doc, value);
       }
     }
-    return Array.from(this._elements);
   }
 
-  getTagNameFromExpression(expression: TagNameExpression):
-      Result<string|undefined, Warning> {
-    if (expression.type === 'string-literal') {
-      return {successful: true, value: expression.value};
+  enterClassExpression(node: estree.ClassExpression, parent: estree.Node) {
+    if (this.alreadyMatched.has(node)) {
+      return;
+    }
+
+    const name = node.id && astValue.getIdentifierName(node.id);
+    const comment = esutil.getAttachedComment(node) ||
+        esutil.getAttachedComment(parent) || '';
+    this._classFound(name, jsdoc.parseJsdoc(comment), node);
+  }
+
+  enterClassDeclaration(node: estree.ClassDeclaration, parent: estree.Node) {
+    const name = astValue.getIdentifierName(node.id);
+    const comment = esutil.getAttachedComment(node) ||
+        esutil.getAttachedComment(parent) || '';
+    this._classFound(name, jsdoc.parseJsdoc(comment), node);
+  }
+
+  private _classFound(
+      name: string|undefined, doc: jsdoc.Annotation, ast: estree.Node) {
+    const namespacedName = name && getNamespacedIdentifier(name, doc);
+
+    this.classes.push({name, namespacedName, doc, ast});
+    if (ast.type === 'ClassExpression') {
+      this.alreadyMatched.add(ast);
     }
-    const element = this._possibleElements.get(expression.className) ||
-        Array.from(this._elements)
-            .find((e) => e.className === expression.className);
-    if (!element) {
-      return {
-        successful: false,
-        value: {
-          code: 'cant-determine-element-tagname',
-          message: `Couldn't dereference the class name ${
-                                                          expression.className
-                                                        } here.`,
-          severity: Severity.WARNING,
-          sourceRange: expression.classNameSourceRange
-        }
-      };
+  }
+}
+
+interface ElementDefineCall {
+  tagName: TagNameExpression;
+  clazz: ElementClassExpression;
+}
+type ElementClassExpression = estree.ClassExpression|{
+  type: 'MaybeChainedIdentifier';
+  name: string, sourceRange: SourceRange
+};
+class CustomElementsDefineCallFinder implements Visitor {
+  readonly warnings: Warning[] = [];
+  readonly calls: ElementDefineCall[] = [];
+  private readonly _document: JavaScriptDocument;
+  constructor(document: JavaScriptDocument) {
+    this._document = document;
+  }
+  enterCallExpression(node: estree.CallExpression) {
+    const callee = astValue.getIdentifierName(node.callee);
+    if (!(callee === 'window.customElements.define' ||
+          callee === 'customElements.define')) {
+      return;
     }
-    return {successful: true, value: element.tagName};
+
+    const tagNameExpression =
+        node.arguments[0] && this._getTagNameExpression(node.arguments[0]);
+    if (tagNameExpression == null) {
```

Comment:
why `tagNameExpression == null` here, and `!elementClassExpression` below? Why not check falsey for both? Please document if there is a specific reason.